client/client-multi/client-c/src/kaa/platform-impl/posix/key_utils.c

```diff
@@ -0,0 +1,417 @@
+#ifdef KAA_RUNTIME_KEY_GENERATION
+            if (rsa_genkey(&pk_context_)) {
+                return -1;
+            }
+#else
+            return KAA_ERR_BADDATA;
+#endif
+        }
+         if (mbedtls_pk_parse_public_keyfile(&pk_pub_context, KAA_PUBLIC_KEY_STORAGE)) {
+#ifdef KAA_RUNTIME_KEY_GENERATION
+		pk_pub_context = pk_context_;
+#else
+            return KAA_ERR_BADDATA;
+#endif
+        }
+        initialized = true;
+    }
+
+    return KAA_ERR_NONE;
+}
+
+void  kaa_deinit_keys(void)
+{
+    mbedtls_pk_free(&pk_context_);
+}
+
+void ext_get_endpoint_session_key(uint8_t **buffer, size_t *buffer_size)
+{
+    if (!buffer || !buffer_size) {
+        return;
+    }
+    *buffer = keys.session_key;
+    *buffer_size = keys.session_key_length;
+}
+
+kaa_error_t ext_encrypt_data(const uint8_t *input, size_t payload_size, uint8_t *output)
+{
+    if (!input || !payload_size || !output) {
+        return KAA_ERR_BADPARAM;
+    }
+
+    /* Adding PKCS7 padding */
+    size_t enc_data_size = ext_get_encrypted_data_size(payload_size);
+    uint8_t padding = enc_data_size - payload_size;
+    memset(output + payload_size, padding, padding);
+
+    /* Process the buffer chunk by chunk */
+    while (enc_data_size >= KAA_SESSION_KEY_LENGTH) {
+        if (aes_encrypt_decrypt(MBEDTLS_AES_ENCRYPT, input,
+                KAA_SESSION_KEY_LENGTH,
+                output, keys.session_key) < 0) {
+            return KAA_ERR_BADDATA;
+        }
+
+        enc_data_size -= KAA_SESSION_KEY_LENGTH;
+
+        input += KAA_SESSION_KEY_LENGTH;
+        output += KAA_SESSION_KEY_LENGTH;
+    }
+
+    return KAA_ERR_NONE;
+}
+
+kaa_error_t ext_decrypt_data(const uint8_t *input, size_t input_size,
+        uint8_t *output, size_t *output_payload_size)
+{
+    if (!input || !input_size || !output || !output_payload_size) {
+        return KAA_ERR_BADPARAM;
+    }
+
+    size_t tmp_in_size = input_size;
+    while (input_size >= KAA_SESSION_KEY_LENGTH) {
+        if (aes_encrypt_decrypt(MBEDTLS_AES_DECRYPT, input,
+                KAA_SESSION_KEY_LENGTH, output,
+                keys.session_key) < 0) {
+            return KAA_ERR_BADDATA;
+        }
+        input_size -= KAA_SESSION_KEY_LENGTH;
+
+        input += KAA_SESSION_KEY_LENGTH;
+        output += KAA_SESSION_KEY_LENGTH;
+    }
+
+    /* Reduce PKCS7 padding */
+    uint8_t padding_length = *(output - 1);
+    *output_payload_size = tmp_in_size - padding_length;
+
+    return KAA_ERR_NONE;
+}
+
+/* Get encrypted session key (enctypted with remote key) */
+kaa_error_t ext_get_encrypted_session_key(uint8_t **buffer, size_t *buffer_size,
+        const uint8_t *remote_key, size_t remote_key_size)
+{
+    if (!buffer || !buffer_size || !remote_key || !remote_key_size) {
+        return KAA_ERR_BADPARAM;
+    }
+
+    mbedtls_pk_context local_ctx;
+    mbedtls_pk_init(&local_ctx);
+    int err = set_rsa_public_key(&local_ctx, remote_key, remote_key_size);
```

Comment:
`set_rsa_public_key` is the Kaa SDK routine. Use error codes from Kaa error namespace.